stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp

```diff
@@ -0,0 +1,174 @@
+ * @param beta weight vector
+ * @param cuts vector of cutpoints
+ * @return log probability
+ * @throw std::domain_error If the outcome is not between 1 and
+ * the number of cutpoints plus 2; or if the cutpoint vector is not sorted in
+ * ascending order; or any input is not finite
+ * @throw std::invalid_argument If y and lambda are different
+ * lengths.
+ * @throw std::invalid_argument if container sizes mismatch.
+ */
+template <bool propto, typename T_y, typename T_x_scalar, typename T_beta,
+          typename T_cuts>
+typename stan::return_type<T_x_scalar, T_beta, T_cuts>::type
+ordered_logistic_glm_lpmf(
+    const T_y& y,
+    const Eigen::Matrix<T_x_scalar, Eigen::Dynamic, Eigen::Dynamic>& x,
+    const T_beta& beta, const T_cuts& cuts) {
+  static const char* function = "ordered_logistic_glm_lpmf";
+  typedef typename partials_return_type<T_y, T_x_scalar, T_beta, T_cuts>::type
+      T_partials_return;
+
+  using Eigen::Array;
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::VectorXd;
+  using std::exp;
+
+  const size_t N_instances = x.rows();
+  const size_t N_attributes = x.cols();
+  const size_t N_classes = length(cuts) + 1;
+
+  check_consistent_size(function, "Vector of dependent variables", y,
+                        N_instances);
+  check_consistent_size(function, "Weight vector", beta, N_attributes);
+  check_bounded(function, "Vector of dependent variables", y, 1, N_classes);
+  check_ordered(function, "Cut-points", cuts);
+  check_finite(function, "Final cut-point", cuts[N_classes - 2]);
+  check_finite(function, "First cut-point", cuts[0]);
+
+  if (size_zero(y, x, beta, cuts))
+    return 0;
+
+  if (!include_summand<propto, T_x_scalar, T_beta, T_cuts>::value)
+    return 0;
+
+  T_partials_return logp(0);
+  const auto& x_val = value_of_rec(x);
+  const auto& beta_val = value_of_rec(beta);
+  const auto& cuts_val = value_of_rec(cuts);
+
+  const auto& y_vec = as_column_vector_or_scalar(y);
+  const auto& beta_val_vec = as_column_vector_or_scalar(beta_val);
+  const auto& cuts_val_vec = as_column_vector_or_scalar(cuts_val);
+
+  Array<double, Dynamic, 1> cuts_y1(N_instances), cuts_y2(N_instances);
+  for (int i = 0; i < N_instances; i++) {
+    int c = y_vec[i];
+    if (c != N_classes) {
+      cuts_y1[i] = cuts_val_vec[c - 1];
+    } else {
+      cuts_y1[i] = INFINITY;
+    }
+    if (c != 1) {
+      cuts_y2[i] = cuts_val_vec[c - 2];
+    } else {
+      cuts_y2[i] = -INFINITY;
+    }
+  }
+
+  Array<double, Dynamic, 1> location = x_val * beta_val_vec;
+  if (!location.allFinite()) {
+    check_finite(function, "Weight vector", beta);
+    check_finite(function, "Matrix of independent variables", x);
+  }
+
+  Array<double, Dynamic, 1> cut2 = location - cuts_y2;
+  Array<double, Dynamic, 1> cut1 = location - cuts_y1;
+
+  // Not immediately evaluating next two expressions benefits performance
+  auto m_log_1p_exp_cut1
+      = -cut1 * (cut1 > 0.0).cast<double>() - (-cut1.abs()).exp().log1p();
```

Comment:
What is the type of `(cut1 > 0)` before the `.cast<double>()`?  I'm having trouble understanding what this cast might be doing.

Reply:
It is `Eigen::Array` of `bool`s. Eigen only allows operations between arrays/matrices that contain same type so we need to cast that to `Eigen::Array` of `double`s

Reply:
I see---it's an Eigen operation---makes sense.  But isn't that particular operation going to be very expensive?  I don't have a good feel for what doing all this multiplication by zero and one for selection is going to cost.  So I think if `cut1 = { -2, 1, 0, -3, 4 }`, then `-cut1 * (cut1 > 0.0).cast<double>()` is just `{ 0, -1, 0, 0, -4 }`.

Reply:
Multiplication is not expensive. I don't think the cast would be either. Anyway I compared this with stan `log_1p_exp()` and it is faster as it is vectorized and contains no branches. I can also try comparing that with `.select()` to see if that would be better.

Reply:
It turns out that `.select()` (which is a branch) is faster.